test/integration/src/main/java/io/pravega/test/integration/selftest/Consumer.java

```diff
@@ -0,0 +1,352 @@
+/**
+ * Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.test.integration.selftest;
+
+import com.google.common.base.Preconditions;
+import io.pravega.common.ExceptionHelpers;
+import io.pravega.common.ObjectClosedException;
+import io.pravega.common.Timer;
+import io.pravega.common.concurrent.CancellationToken;
+import io.pravega.common.concurrent.FutureHelpers;
+import io.pravega.common.util.BlockingDrainingQueue;
+import io.pravega.segmentstore.contracts.StreamSegmentSealedException;
+import io.pravega.test.integration.selftest.adapters.StoreAdapter;
+import io.pravega.test.integration.selftest.adapters.StoreReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.time.Duration;
+import java.util.AbstractMap;
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Map;
+import java.util.Queue;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ScheduledExecutorService;
+import javax.annotation.concurrent.GuardedBy;
+import lombok.SneakyThrows;
+import lombok.val;
+
+/**
+ * Represents an operation consumer. Attaches to a StoreAdapter, listens to tail-reads (on segments), and validates
+ * incoming data. Currently this does not do catch-up reads or storage reads/validations.
+ */
+public class Consumer extends Actor {
```

Comment:
There are so many changes in this file I think it's best if you don't try a diff. I mostly rewrote the whole thing.